apiserver/facades/client/cloud/cloud_test.go

```diff
@@ -1037,206 +1368,270 @@ func (s *cloudSuite) TestRevokeCredentials(c *gc.C) {
+	}
+	for _, l := range expectedLog {
+		c.Assert(c.GetTestLog(), jc.Contains, l)
 	}
-	s.assertRevokeCredentials(c, t)
 }
 
 func (s *cloudSuite) TestRevokeCredentialsClearModelCredentialsError(c *gc.C) {
-	s.backend.SetErrors(
-		nil,                  // RemoveCloudCredential
-		errors.New("kaboom"), // RemoveModelsCredential
-	)
-	t := revokeCredentialData{
-		f: func(tag names.CloudCredentialTag) (map[string]string, error) {
-			return map[string]string{
-				coretesting.ModelTag.Id(): "modelName",
-			}, nil
-		},
-		args: []params.RevokeCredentialArg{
-			{Tag: "cloudcred-meep_julia_three", Force: true},
-		},
-		callsMade: []string{"ControllerTag", "CredentialModels", "RemoveCloudCredential", "RemoveModelsCredential"},
-		results: params.ErrorResults{
-			Results: []params.ErrorResult{
-				{apiservererrors.ServerError(errors.New("kaboom"))},
-			},
-		},
-		expectedLog: []string{
-			` WARNING juju.apiserver.cloud credential cloudcred-meep_julia_three will be deleted but it is used by model deadbeef-0bad-400d-8000-4b1d0d06f00d`,
+	adminTag := names.NewUserTag("admin")
+	defer s.setup(c, adminTag).Finish()
+
+	model := map[string]string{
+		"deadbeef-0bad-400d-8000-4b1d0d06f00d": "testModel",
+	}
+
+	tag, err := cloudCredentialTag("three", "julia", "meep", map[string]string{}).CloudCredentialTag()
+	c.Assert(err, jc.ErrorIsNil)
+
+	backend := s.backend.EXPECT()
+	backend.CredentialModels(tag).Return(model, nil)
+	backend.RemoveCloudCredential(tag).Return(nil)
+	backend.RemoveModelsCredential(tag).Return(errors.New("kaboom"))
+
+	results, err := s.api.RevokeCredentialsCheckModels(params.RevokeCredentialArgs{[]params.RevokeCredentialArg{
+		{Tag: "cloudcred-meep_julia_three", Force: true},
+	}})
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(results, gc.DeepEquals, params.ErrorResults{
+		Results: []params.ErrorResult{
+			{apiservererrors.ServerError(errors.New("kaboom"))},
 		},
+	})
+	expectedLog := []string{
+		` WARNING juju.apiserver.cloud credential cloudcred-meep_julia_three will be deleted but it is used by model deadbeef-0bad-400d-8000-4b1d0d06f00d`,
+	}
+	for _, l := range expectedLog {
+		c.Assert(c.GetTestLog(), jc.Contains, l)
 	}
-	s.assertRevokeCredentials(c, t)
 }
 
 func (s *cloudSuite) TestCredential(c *gc.C) {
-	s.setTestAPIForUser(c, names.NewUserTag("bruce"))
+	bruceTag := names.NewUserTag("bruce")
+	defer s.setup(c, bruceTag).Finish()
+
+	tagOne, err := cloudCredentialTag("foo", "admin", "meep", map[string]string{}).CloudCredentialTag()
+	c.Assert(err, jc.ErrorIsNil)
+	tagTwo, err := cloudCredentialTag("two", "bruce", "meep", map[string]string{}).CloudCredentialTag()
+	c.Assert(err, jc.ErrorIsNil)
+
+	creds := map[string]state.Credential{
+		tagOne.Id(): statetesting.NewEmptyCredential(),
+		tagTwo.Id(): statetesting.CloudCredential(jujucloud.UserPassAuthType, map[string]string{
+			"username": "admin",
+			"password": "adm1n",
+		}),
+	}
```

Comment:
For some reason as soon as you generate the credentials that should be used here, you drop them in favour of their tag and the reconstruct the credential manually (and in an incomplete fashion).

Instead, I would generate both the credential and the tag in `cloudCredentialTag`. This would mean you can drop the error check every time you create the tag in your tests by doing it inside `cloudCrednetialTag`

Reply:
Refactored the `cloudCrendentialTag` as described in the below comment.